### PR TITLE
[FIRRTL] Add support for "intrinsics"

### PR DESCRIPTION
**Pull request by darthscsi**
Add support for intrinsics in FIRRTL.  Until intrinsics are supported in the firrtl spec, or we have more complex intrinsics, do direct lowering of intrinsics into their ops.

It is expected that eventually something like an "intmodule" will exist and annotated extern modules will not be used for intrinsics or first they will lower to intmodules internally, then intmodules will be handled uniformly.

As an example, add a sizeof operator which returns the number of bits in the argument type.  This let's you query the result of type inference from inside the circuit.

### CODE REVIEW

File: tools/firtool/firtool.cpp [hunk 1]
```diff
@@ -612,6 +612,8 @@ processBuffer(MLIRContext &context, TimingScope &ts, llvm::SourceMgr &sourceMgr,
     return success();
   }
 
+  pm.nest<firrtl::CircuitOp>().addPass(firrtl::createLowerIntrinsicsPass());
```
Comment (mikeurbach) [inline tools/firtool/firtool.cpp:615]: Is it worth running this for `--parse-only` like lower annotations? Might be handy if you're just testing new intrinsics, but it isn't exactly "only parsing" at that point.

File: lib/Dialect/FIRRTL/Transforms/LowerIntrinsics.cpp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+      }
+      if (!ports[0].getName().equals("i")) {
+        mod.emitError("circt.sizeof first port named '")
+            << ports[0].getName() << "' instead of 'i'";
+        return false;
+      }
+      if (!ports[1].getName().equals("size")) {
+        mod.emitError("circt.sizeof second port named '")
+            << ports[0].getName() << "' instead of 'size'";
+return false;
```
Comment (mikeurbach) [inline lib/Dialect/FIRRTL/Transforms/LowerIntrinsics.cpp:54]: Seems some of the returns in this file are not indented.

File: lib/Dialect/FIRRTL/Transforms/LowerIntrinsics.cpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    if (intname.getValue().equals("circt.sizeof")) {
+      if (lowerCirctSizeof(instancePathCache, op)) {
+        ++numConverted;      
+        op.erase();
+      } else {
+      ++numFailures;
+      }
+      continue;
+    }
+        if (intname.getValue().equals("circt.isX")) {
```
Comment (mikeurbach) [inline lib/Dialect/FIRRTL/Transforms/LowerIntrinsics.cpp:150]: Indentation looks weird here too.

File: lib/Dialect/FIRRTL/Transforms/LowerIntrinsics.cpp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+namespace {
+struct LowerIntrinsicsPass : public LowerIntrinsicsBase<LowerIntrinsicsPass> {
+  void runOnOperation() override;
+};
+} // end anonymous namespace
+
+static bool lowerCirctSizeof(InstancePathCache& instancePathCache, FExtModuleOp mod) {
+      auto ports = mod.getPorts();
+      if (ports.size() != 2) {
```
Comment (mikeurbach) [inline lib/Dialect/FIRRTL/Transforms/LowerIntrinsics.cpp:42]: I wish there was a better way to separate verification from transformation, but I can't think of a better way for the short term implementation we have here...

File: lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp [hunk 5]
```diff
@@ -294,9 +294,9 @@ struct AnnoRecord {
 /// the FIRRTL Circuit, i.e., an Annotation which has no target.  Historically,
 /// NoTargetAnnotations were used to control the Scala FIRRTL Compiler (SFC) or
 /// its passes, e.g., to set the output directory or to turn on a pass.
-/// Examplesof these in the SFC are "firrtl.options.TargetDirAnnotation" to set
+/// Examples of these in the SFC are "firrtl.options.TargetDirAnnotation" to set
 /// the output directory or "firrtl.stage.RunFIRRTLTransformAnnotation" to
-/// casuse the SFC to schedule a specified pass.  Instead of leaving these
+/// caause the SFC to schedule a specified pass.  Instead of leaving these
```
Comment (mikeurbach) [inline lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp:299]: nit: from one typo to another

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp [hunk 6]
```diff
@@ -888,6 +888,14 @@ LogicalResult NEQPrimOp::canonicalize(NEQPrimOp op, PatternRewriter &rewriter) {
 // Unary Operators
 //===----------------------------------------------------------------------===//
 
+OpFoldResult SizeOfIntrinsicOp::fold(llvm::ArrayRef<mlir::Attribute>) {
+  auto base = getInput().getType().dyn_cast<FIRRTLBaseType>();
```
Comment (mikeurbach) [inline lib/Dialect/FIRRTL/FIRRTLFolds.cpp:892]: nit: i think either `base` should be checked for null, or else this can be a `cast`.

**Review by mikeurbach — COMMENTED** [review]
Nice, left some minor comments while I was looking through it.

File: lib/Dialect/FIRRTL/Transforms/LowerIntrinsics.cpp (hunk 4 above)
Comment (darthscsi) [reply lib/Dialect/FIRRTL/Transforms/LowerIntrinsics.cpp]: There's some table-driven verification possible.

Comment (darthscsi) [reply lib/Dialect/FIRRTL/Transforms/LowerIntrinsics.cpp]: And we can fake it with some factoring.

**Pull request merged**